### PR TITLE
Debounce

### PR DESCRIPTION
**Pull request by NouraA11**
This is for merging changes like: 
-Code refactoring
-Fix infinite loading function
-Fix debounce function

### CODE REVIEW

File: src/components/SearchBar/SearchBar.style.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { StyleSheet } from "react-native";
+
+export default StyleSheet.create({
+    searchBarContainer: {
+        marginBottom: 10,
+        backgroundColor: "#ffffff",
+        borderColor: 'transparent',
```
Comment (rfziade) [inline src/components/SearchBar/SearchBar.style.js:7]: do we need this?

File: src/components/NavigationButton/NavigationButton.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import React from "react";
+import { useNavigation } from "@react-navigation/native";
+import { TouchableOpacity } from 'react-native';
+import { Icon } from "react-native-elements";
+
+const NavigationButton = () => {
+    const navigation = useNavigation();
+
+    const handleClick = () => {
+        navigation.navigate('Favorites');
```
Comment (rfziade) [inline src/components/NavigationButton/NavigationButton.js]: can we have screenNames in a global file of their own please?

File: src/components/RepoList/RepoList.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import {FlatList, ActivityIndicator} from 'react-native';
+import Card from '../../components/Card/Card';
+
+const renderItem = ({ item }) => <Card repo={item} />
+
+const RepoList = ({ query, initialDataItems, searchDataItems, isFetchingNextPage, handleLoadMore }) => {
+
+    return (
+        <FlatList
+            keyExtractor={(item, index) => `${index}-${item.id}`}
```
Comment (rfziade) [inline src/components/RepoList/RepoList.js]: keyExtractor function outside the component

File: src/components/SearchBar/SearchBar.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        if (value.replace(/[^\w_@./#&+-]*$/, "") !== "") {
+            setSearchValue(value)
+
+            // debounce(onChangeText, 2000)(value)
+            debouncedFunction(value);
+        }
+    }
+            
+    const handleClear = () => {
+        onChangeText('')
```
Comment (rfziade) [inline src/components/SearchBar/SearchBar.js:23]: Shouldn't we change the search value too?

File: src/hooks/useRepos.js [hunk 5]
```diff
@@ -1,18 +1,20 @@
...
-import Config from "react-native-config";
+import { queryKeysList } from "../utils/constants";
+import { INITIAL_DATA_PER_PAGE, SORT_BY } from "../utils/config";
+import instance from "../utils/instance";
 
 //returns popular repositories to display as initial data
 const fetchRepos = () => {
-    axios.defaults.headers.common['Authorization'] = Config.GITHUB_ACCESS_TOKEN;
-    return axios.get('https://api.github.com/repositories?per_page=70&sort=stars')
+    return instance.get(`/repositories?per_page=${INITIAL_DATA_PER_PAGE}&sort=${SORT_BY}`);
```
Comment (rfziade) [inline src/hooks/useRepos.js]: please fix params

File: src/hooks/useRepos.js [hunk 6]
```diff
@@ -1,18 +1,20 @@
...
 }
 
 const useRepos = (query, onSuccess, onError) => {
+    const {INITIAL_REPOS_KEY} = queryKeysList;
```
Comment (rfziade) [inline src/hooks/useRepos.js]: outside hook please

File: src/hooks/useSearch.js [hunk 7]
```diff
@@ -1,26 +1,28 @@
...
 import { useInfiniteQuery } from "react-query";
+import { queryKeysList } from "../utils/constants";
+import { SEARCH_DATA_PER_PAGE } from "../utils/config";
+import instance from "../utils/instance";
 
 const fetchRepos = ({ pageParam = 1, queryKey }) => {
     const [, query] = queryKey;
-    axios.defaults.headers.common['Authorization'] = Config.GITHUB_ACCESS_TOKEN;
-    return axios.get(`https://api.github.com/search/repositories?q=${query}&per_page=20&page=${pageParam}`)
+    return instance.get(`/search/repositories?q=${query}&per_page=${SEARCH_DATA_PER_PAGE}&page=${pageParam}`)
```
Comment (rfziade) [inline src/hooks/useSearch.js]: query params please

File: src/navigation/NavigationComponent.js [hunk 8]
```diff
@@ -1,25 +1,28 @@
...
-import NavigationButton from "../components/NavigationButton";
+import Home from "../screens/home/Home";
+import Favorites from "../screens/favorites/Favorites";
+import NavigationButton from "../components/NavigationButton/NavigationButton";
+import {screenNames} from '../utils/constants'
 
 const Stack = createNativeStackNavigator();
 
 const NavigationComponent = () => {
+    const {home, favorite} = screenNames;
```
Comment (rfziade) [inline src/navigation/NavigationComponent.js]: always when having values independent from renders such as this case,
Always put them outside the component please

File: src/screens/Favorites/Favorites.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import React from "react";
+import { Text, View, ScrollView } from "react-native";
+import { useSelector } from "react-redux";
+import styles from './Favorites.style';
+import Card from "../../components/Card/Card";
+
+const Favorites = () => {
+    const favoriteList = useSelector((state) => state.Favorite);
+    
+    const FavoriteListComponent = () => (
```
Comment (rfziade) [inline src/screens/Favorites/Favorites.js]: having components inside components is conventionally not recommended but it is not wrong,
Usually tho, we either move the components outside the component's code
Or we rename to "renderFavoriteListComponent" and we use it "renderFavoriteListComponent()"

But that is just conventions and has negligible impact on performance or readabiltiy

File: src/screens/Home/Home.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    const handleLoadMore = () => {
+        if (hasNextPage && !isFetchingNextPage) {
+            fetchNextPage();
+        }
+    }
+
+    const initialDataItems = initialData?.data;
+    const searchDataItems = searchData?.pages?.map(page => page.data.items).flat();
+
+    const RepoListProps = {
```
Comment (rfziade) [inline src/screens/Home/Home.js]: variables start with lowercase please

File: src/screens/Home/Home.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    const onSuccess = (data) => {
+        console.log('Successfully fetched')
+    }
+    const onError = (error) => {
+        console.log('Error encountered ', error.message)
+    }
+
+    const { data: initialData, isLoading: isLoadingInitial, isError: isErrorInitial } = useRepos(query, onSuccess, onError)
+    const { data: searchData, isLoading: isLoadingSearch, isError: isErrorSearch, 
+        refetch, fetchNextPage, hasNextPage, isFetchingNextPage, status } = useSearch(query, onSuccess, onError);
```
Comment (rfziade) [inline src/screens/Home/Home.js:21]: We need to discuss the API calls on query change please when possible.
This is something I missed to tell you about before and I just noticed it.

**Review by rfziade — CHANGES_REQUESTED** [review]

File: src/screens/Home/Home.js (hunk 11 above)
Comment (rfziade) [reply src/screens/Home/Home.js:21]: As discussed, nothing to do here, working fine.
Thanks.

File: src/components/SearchBar/SearchBar.style.js (hunk 1 above)
Comment (NouraA11) [reply src/components/SearchBar/SearchBar.style.js:7]: Yes, it's needed

File: src/components/NavigationButton/NavigationButton.js (hunk 2 above)
Comment (NouraA11) [reply src/components/NavigationButton/NavigationButton.js]: screenNames is already in a global file, however, using it was missed here. Now it's fixed.

File: src/components/RepoList/RepoList.js (hunk 3 above)
Comment (NouraA11) [reply src/components/RepoList/RepoList.js]: keyExtractor function is extracted now.

File: src/components/SearchBar/SearchBar.js (hunk 4 above)
Comment (NouraA11) [reply src/components/SearchBar/SearchBar.js:23]: Done

File: src/hooks/useRepos.js (hunk 6 above)
Comment (NouraA11) [reply src/hooks/useRepos.js]: It was fixed as we discussed.

File: src/screens/Home/Home.js (hunk 10 above)
Comment (NouraA11) [reply src/screens/Home/Home.js]: Done.

File: src/screens/Favorites/Favorites.js (hunk 9 above)
Comment (NouraA11) [reply src/screens/Favorites/Favorites.js]: I agree, so I moved them outside the component code.

File: src/navigation/NavigationComponent.js (hunk 8 above)
Comment (NouraA11) [reply src/navigation/NavigationComponent.js]: It is fixed now.

File: src/hooks/useRepos.js (hunk 5 above)
Comment (NouraA11) [reply src/hooks/useRepos.js]: As we discussed, I replaced it with a params object, since it's a better way of implementing it.

**Review by rfziade — APPROVED** [review]

**Pull request merged**